File: worker/src/calendar.js
// calendar.js — the season skeleton ESPN ships inside every scoreboard's
// leagues[0]. Pure (no I/O) so it runs identically in Node tests and the Worker,
// and it is the SINGLE home for reading ESPN's calendar (overview.js's
// season-pulse classifier and normalize.js's scores passthrough both import it —
// the "interpret a league's calendar" rule must never fork).
//
// Two shapes by `calendarType`:
//   "day"  → calendar is an ISO date[] (one entry per GAME DAY): NBA/NHL/MLB and
//            soccer leagues. This IS the authoritative "which days have games"
//            list — exactly what the league-detail Schedule strip needs.
//   "list" → calendar is an object[] of season-type buckets / events with
//            [startDate,endDate] ranges (NFL nests weeks under `entries`;
//            golf/F1/MMA list one event per entry). NOT per-day game presence.

// US-Eastern calendar day as a UTC-midnight stamp, matching ESPN's bucketing (and
// the app's "today"). Intl resolves EST/EDT automatically. Null if unparsable.
export function easternDayMs(input) {
  const d = input instanceof Date ? input : new Date(input);
  if (Number.isNaN(d.getTime())) return null;
  const parts = new Intl.DateTimeFormat('en-US', {
    timeZone: 'America/New_York', year: 'numeric', month: '2-digit', day: '2-digit',
  }).formatToParts(d);
  const g = (t) => Number(parts.find((p) => p.type === t).value);
  return Date.UTC(g('year'), g('month') - 1, g('day'));
}

// Collapse a league's calendar into sorted ranges (ET days), each tagged with
// whether it came from a NESTED season bucket (a week/phase under `entries`) vs a
// flat top-level entry (a single event). Shape: { start, end, nested }.
export function rangesFromCalendar(calendarType, calendar) {
  const out = [];
  if (!Array.isArray(calendar) || !calendar.length) return out;
  const isDay = calendarType === 'day' || typeof calendar[0] === 'string';
  if (isDay) {
    for (const s of calendar) {
      const ms = easternDayMs(s);
      if (ms != null) out.push({ start: ms, end: ms, nested: false });
    }
  } else {
    for (const entry of calendar) {
      // NFL/UFL nest weeks and soccer nests competition phases under `entries`;
      // golf/F1/MMA are one event per (flat) entry. The nested children are
      // season buckets, NOT individual game days.
      const nested = Array.isArray(entry.entries) && entry.entries.length > 0;
      const kids = nested ? entry.entries : [entry];
      for (const k of kids) {
        if (!k || !k.startDate) continue;
        const start = easternDayMs(k.startDate);
        if (start == null) continue;
        // Pull the end back 1s so an end stamped at ET-midnight doesn't bleed
        // into the following day.
        const end = k.endDate ? easternDayMs(new Date(new Date(k.endDate).getTime() - 1000)) : start;
        out.push({ start, end: Math.max(end ?? start, start), nested });
      }
    }
  }
  return out.sort((a, b) => a.start - b.start);
}

// 'YYYYMMDD' for an ET-day stamp (the app's date-param + day-key format).
const pad = (n) => (n < 10 ? '0' : '') + n;
export const ymd = (ms) => {
  const d = new Date(ms);
  return `${d.getUTCFullYear()}${pad(d.getUTCMonth() + 1)}${pad(d.getUTCDate())}`;
};

// The on-the-scores-payload calendar: a precise game-day list for "day"-type
// leagues (drives the Schedule strip's empty-day dimming + auto-focus WITHOUT a
// separate range fetch — the data already rides the scoreboard we fetched), plus
// the season window for the offseason-opener jump. Returns {} when absent.
// "list"-type (gridiron/golf/F1/MMA) is deliberately omitted: its ranges are
// week/season buckets, not per-day game presence, so the app keeps its
// event-derived strip there (and gridiron games already carry weekLabel).
export function buildCalendar(lg) {
  const out = {};
  const cal = lg && lg.calendar;
  const isDay = lg && (lg.calendarType === 'day' || (Array.isArray(cal) && typeof cal[0] === 'string'));
  if (isDay && Array.isArray(cal) && cal.length) {
    const set = new Set();
    for (const s of cal) {
      const ms = easternDayMs(s);
      if (ms != null) set.add(ymd(ms));
    }
    if (set.size) out.calendarDays = [...set].sort();
  }
  const season = (lg && lg.season) || {};
  const win = {};
  if (season.startDate) win.startDate = season.startDate;
  if (season.endDate) win.endDate = season.endDate;
  if (win.startDate || win.endDate) out.seasonWindow = win;
  return out;
}

File: app/lib/src/data/calendar.dart
// calendar.dart — Dart port of worker/src/calendar.js. The season skeleton ESPN
// ships inside every scoreboard's leagues[0]. Pure. Single home for reading
// ESPN's calendar (normalize.dart's scores passthrough + overview.dart's
// season-pulse classifier both use it — the rule must never fork).
//
// The JS reads US-Eastern calendar days via Intl.DateTimeFormat('America/
// New_York'). Dart has no bundled IANA tz DB, so we compute the US Eastern
// offset from the standard DST rules (EDT = UTC-4 from the 2nd Sunday of March
// 02:00 to the 1st Sunday of November 02:00; EST = UTC-5 otherwise). This matches
// Intl for every US Eastern date — verified against the JS output via the scores
// golden suite (calendarDays parity).

/// nth Sunday of a month as a UTC instant at [utcHour]:00.
int _nthSundayUtc(int year, int month, int nth, int utcHour) {
  var d = DateTime.utc(year, month, 1);
  var count = 0;
  while (true) {
    if (d.weekday == DateTime.sunday) {
      count++;
      if (count == nth) break;
    }
    d = d.add(const Duration(days: 1));
  }
  return DateTime.utc(year, month, d.day, utcHour).millisecondsSinceEpoch;
}

/// US Eastern UTC offset (in hours, negative) for a given UTC instant.
int _etOffsetHours(int utcMs) {
  final year = DateTime.fromMillisecondsSinceEpoch(utcMs, isUtc: true).year;
  // DST spring-forward: 2nd Sunday of March at 02:00 EST == 07:00 UTC.
  final dstStart = _nthSundayUtc(year, 3, 2, 7);
  // DST fall-back: 1st Sunday of November at 02:00 EDT == 06:00 UTC.
  final dstEnd = _nthSundayUtc(year, 11, 1, 6);
  return (utcMs >= dstStart && utcMs < dstEnd) ? -4 : -5;
}

/// US-Eastern calendar day as a UTC-midnight stamp, matching ESPN's bucketing
/// (and the app's "today"). Returns null if unparsable.
int? easternDayMs(dynamic input) {
  int? utcMs;
  if (input is int) {
    utcMs = input;
  } else if (input is DateTime) {
    utcMs = input.millisecondsSinceEpoch;
  } else if (input is String) {
    final d = DateTime.tryParse(input);
    if (d != null) utcMs = d.millisecondsSinceEpoch;
  }
  if (utcMs == null) return null;
  final local = DateTime.fromMillisecondsSinceEpoch(
      utcMs + _etOffsetHours(utcMs) * 3600 * 1000,
      isUtc: true);
  return DateTime.utc(local.year, local.month, local.day).millisecondsSinceEpoch;
}

String _pad(int n) => n < 10 ? '0$n' : '$n';

/// 'YYYYMMDD' for an ET-day stamp (the app's date-param + day-key format).
String ymd(int ms) {
  final d = DateTime.fromMillisecondsSinceEpoch(ms, isUtc: true);
  return '${d.year}${_pad(d.month)}${_pad(d.day)}';
}

/// Collapse a league's calendar into sorted ranges (ET days), each tagged with
/// whether it came from a NESTED season bucket vs a flat top-level entry.
/// Shape: { start, end, nested }. Mirrors rangesFromCalendar in calendar.js.
List<Map<String, dynamic>> rangesFromCalendar(dynamic calendarType, dynamic calendar) {
  final out = <Map<String, dynamic>>[];
  if (calendar is! List || calendar.isEmpty) return out;
  final isDay = calendarType == 'day' || calendar[0] is String;
  if (isDay) {
    for (final s in calendar) {
      final ms = easternDayMs(s);
      if (ms != null) out.add({'start': ms, 'end': ms, 'nested': false});
    }
  } else {
    for (final entry in calendar) {
      if (entry is! Map) continue;
      final entries = entry['entries'];
      final nested = entries is List && entries.isNotEmpty;
      final kids = nested ? entries : [entry];
      for (final k in kids) {
        if (k is! Map || k['startDate'] == null) continue;
        final start = easternDayMs(k['startDate']);
        if (start == null) continue;
        final endRaw = k['endDate'];
        int? end = start;
        if (endRaw != null) {
          final parsed = DateTime.tryParse(endRaw.toString());
          if (parsed != null) {
            end = easternDayMs(parsed.millisecondsSinceEpoch - 1000);
          }
        }
        out.add({'start': start, 'end': (end ?? start) > start ? end : start, 'nested': nested});
      }
    }
  }
  out.sort((a, b) => (a['start'] as int).compareTo(b['start'] as int));
  return out;
}

/// The on-the-scores-payload calendar: a precise game-day list for "day"-type
/// leagues + the season window. Returns {} when absent. Mirrors buildCalendar.
Map<String, dynamic> buildCalendar(dynamic lg) {
  final out = <String, dynamic>{};
  final cal = lg is Map ? lg['calendar'] : null;
  final isDay = lg is Map &&
      (lg['calendarType'] == 'day' || (cal is List && cal.isNotEmpty && cal[0] is String));
  if (isDay && cal is List && cal.isNotEmpty) {
    final set = <String>{};
    for (final s in cal) {
      final ms = easternDayMs(s);
      if (ms != null) set.add(ymd(ms));
    }
    if (set.isNotEmpty) {
      final sorted = set.toList()..sort();
      out['calendarDays'] = sorted;
    }
  }
  final season = (lg is Map ? lg['season'] : null) ?? {};
  final win = <String, dynamic>{};
  if (season is Map && season['startDate'] != null) win['startDate'] = season['startDate'];
  if (season is Map && season['endDate'] != null) win['endDate'] = season['endDate'];
  if (win.isNotEmpty) out['seasonWindow'] = win;
  return out;
}
